java/server/src/org/openqa/grid/common/SeleniumProtocol.java

```diff
@@ -17,8 +17,44 @@
 
 package org.openqa.grid.common;
 
+import static org.openqa.grid.common.RegistrationRequest.PATH;
+import static org.openqa.grid.common.RegistrationRequest.SELENIUM_PROTOCOL;
+
+import org.openqa.grid.common.exception.GridException;
+import org.openqa.selenium.remote.DesiredCapabilities;
+
+import java.util.Arrays;
+import java.util.Map;
+
 public enum SeleniumProtocol {
-  Selenium, WebDriver;
+  Selenium("/selenium-server/driver"),
+  WebDriver("/wd/hub");
+  private String path;
+
+  SeleniumProtocol(String path) {
+    this.path = path;
+  }
+
+  public static SeleniumProtocol getProtocol(DesiredCapabilities capability) {
+    String type = (String) capability.getCapability(SELENIUM_PROTOCOL);
+    if (type == null || type.trim().isEmpty()) {
+      return WebDriver;
+    }
+    try {
+      return SeleniumProtocol.valueOf(type);
+    } catch (IllegalArgumentException e) {
+      throw new GridException(type + " isn't a valid protocol type for grid. Valid values :[" +
+                              Arrays.toString(values()) + "]", e);
+    }
+  }
+
+  public String getPath(Map<String, Object> capabilities) {
```

Comment:
It seems inconsistent to me that `getProtocol` takes `DesiredCapabilities` and `getPath` takes a `Map<String, Object>`.   Can they both take `DesiredCapabilities` as the original `private` code (in `BaseRemoteProxy`) that this was moved from did? 

Also, now that these methods are `public`, I'd like to suggest names of 
- `public static SeleniumProtocol  fromDesiredCapabilities()`
- `public String getPathConsideringDesiredCapabilities()`
- `public String getPath()` which returns ONLY the configured (via the enum's constructor) `this.path`